pkg/maglev/maglev.go

```diff
@@ -60,16 +62,35 @@ func getOffsetAndSkip(backend string, m uint64) (uint64, uint64) {
 	return offset, skip
 }
 
-func getPermutation(backends []string, m uint64) []uint64 {
-	perm := make([]uint64, len(backends)*int(m))
+func getPermutation(backends []string, m uint64, numCPU int) []uint64 {
+	var wg sync.WaitGroup
+	bCount := len(backends)
+	perm := make([]uint64, bCount*int(m))
 
-	for i, backend := range backends {
-		offset, skip := getOffsetAndSkip(backend, m)
-		perm[i*int(m)] = offset % m
-		for j := uint64(1); j < m; j++ {
-			perm[i*int(m)+int(j)] = (perm[i*int(m)+int(j-1)] + skip) % m
-		}
+	batchSize := bCount / numCPU
+	if batchSize == 0 {
+		batchSize = bCount
+	}
+
+	for g := 0; g < bCount; g += batchSize {
+		wg.Add(1)
+		go func(from int) {
+			to := from + batchSize
+			if to > bCount {
+				to = bCount
+			}
+			for i := from; i < to; i++ {
+				offset, skip := getOffsetAndSkip(backends[i], m)
+				perm[i*int(m)] = offset % m
+				for j := uint64(1); j < m; j++ {
+					perm[i*int(m)+int(j)] = (perm[i*int(m)+int(j-1)] + skip) % m
+				}
+
+			}
+			wg.Done()
+		}(g)
```

Comment:
As mentioned in an offline discussion, using a semaphore here would constrain the number of goroutines as well. However, there's a mention of wanting to avoid creating a goroutine per backend, which can be expensive given a large number of backends. This would presumably put pressure on the Golang scheduler to schedule a potentially large number of goroutines, even if short lived, as well as taking up a much larger footprint of memory.

From my quick reading of the code, this seems like it's splitting the "work" into chunks and having a goroutine process the chunk. I think this is a preferable approach as it would only create up to X goroutines, rather than `len(backends)`.

Reply:
> The optimization gives ~5x improvement in time:

Nice! Did you also assess the overhead / time for smaller number of backends and/or M? Mainly wondering if the heuristic should be adapted to only kick in when the overhead of spawning extra go routines brings a measurable win ... otoh I don't think people would configure very small M.

Reply:
With smaller Ms I got 4x.